src/lpm.c

```diff
@@ -711,110 +718,233 @@ static int lpm_extract(lua_State* L) {
+        
       mtar_header_t h;
+      mtar_header_t before_h;
+      mtar_header_t allways_h;
+      int has_ext_before = 0;
+      int has_ext_allways = 0;
+      
       while ((mtar_read_header(&tar, &h)) != MTAR_ENULLRECORD ) {
         if (h.type == MTAR_TREG) {
+        
+          if (has_ext_before) {
+            mtar_update_header(&h, &before_h);
+            has_ext_before = 0;
+            mtar_clear_header(&before_h);
+          }
+          if (has_ext_allways)
+            mtar_update_header(&h, &allways_h);
+          
           char target[MAX_PATH];
           int target_length = snprintf(target, sizeof(target), "%s/%s", dst, h.name);
+          
           if (mkdirp(target, target_length)) {
             mtar_close(&tar);
             return luaL_error(L, "can't extract tar archive file %s, can't create directory %s: %s", src, target, strerror(errno));
           }
+          
           char buffer[8192];
           FILE* file = fopen(target, "wb");
           if (!file) {
             mtar_close(&tar);
             return luaL_error(L, "can't extract tar archive file %s, can't create file %s: %s", src, target, strerror(errno));
           }
+          
           if (chmod(target, h.mode))
             return luaL_error(L, "can't extract tar archive file %s, can't chmod file %s: %s", src, target, strerror(errno));
+            
           int remaining = h.size;
           while (remaining > 0) {
             int read_size = remaining < sizeof(buffer) ? remaining : sizeof(buffer);
+            
             if (mtar_read_data(&tar, buffer, read_size) != MTAR_ESUCCESS) {
               fclose(file);
               mtar_close(&tar);
               return luaL_error(L, "can't write file %s: %s", target, strerror(errno));
             }
+            
             fwrite(buffer, sizeof(char), read_size, file);
             remaining -= read_size;
           }
+          
           fclose(file);
         }
+        
+        // !!! NOT REALLY TESTED !!!
+        else if (h.type == MTAR_TEHR || h.type == MTAR_TEHRA) {
+          mtar_header_t *h_to_change;
+          if (h.type == MTAR_TEHR)
+            h_to_change = &before_h;
+          else
+            h_to_change = &allways_h;
+          
+          char buffer[4096];
+          char curent_read[8192]; // If a line is more than 8192 char long, will not work!
+          char last_read[4096];
+          int remaining = h.size;
+          
+          
+          has_ext_before = 1;
+          
+          while (remaining > 0) {
+            int read_size = remaining < sizeof(buffer) ? remaining : sizeof(buffer);
+            remaining -= read_size;
+            
+            if (mtar_read_data(&tar, buffer, read_size) != MTAR_ESUCCESS) {
+              mtar_close(&tar);
+              return luaL_error(L, "Error while reading extended: %s", strerror(errno));
+            }
+            
+            strcpy(curent_read, last_read);
```

Comment:
I may be wrong, but on the first run, `last_read` is initialized (l.879), but does not contains anything yet (which is normal, because it is the first run). I don't see where is the problem here.

Reply:
I guess my point is (and I could be wrong here); is it not possible that `last_read` contains arbitrary garbage data on the read, unless you declare it as `char last_read[4096]={0}`?

Reply:
Well all I know is that I am not good enough in C to say that, so I'm going to fully trust you on this.

Reply:
Honestly, your code is great regardless. But yeah; I think it'd be better. Do you have a tar archive which would trigger this branch, so I can test?

Reply:
I don't have one for now, but it should be easy to generate.

Reply:
Hrmm. OK, yeah; I looked it up, and generated one with `tar`. Upon extraction, it goes into an infinite loop.

Also, yeah; I checked with valgrind, and it does require the initialization, otherwise it gets flagged as an uninitialized access, which is UB.

Also, because of the `strcat` on line 895; we also need to initialize `buffer` and `curent_read` to 0 (also, I'd say add an `r` to that so it's `current_read`).

Reply:
Ok, I'll fix all of this tomorrow. In the mean time, can I have your `tar` test file so I will be sure to run into the same bugs as you?

Reply:
Unfortunately github doesn't let me upload tars, but you can reproduce in the repo working directory by doing:

```
tar --format=pax -cvf test.tar README.md && ./lpm extract test.tar
```

Reply:
Ok, it seems to work now, I've tested it with a file with a very long name also.

Reply:
Looks good!